# Рекомендательная система нового тарифа на основе данных о поведении абонентов

Оператор мобильной связи «Мегалайн» выяснил, что многие клиенты пользуются архивными тарифами. Необходимо построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».
Для этих целей есть данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится, она была сделана в предыдущей работе.
Необходимо построить модель с максимально большим значением accuracy. Требуется довести долю правильных ответов по крайней мере до 0.75. Необходимо проверить accuracy на тестовой выборке.

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Инструкция-по-выполнению-проекта" data-toc-modified-id="Инструкция-по-выполнению-проекта-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Инструкция по выполнению проекта</a></span></li></ul></li><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Считывание-данных" data-toc-modified-id="Считывание-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Считывание данных</a></span></li><li><span><a href="#Разделение-данных-на-обучающую,-валидационную-и-тестовую-выборки." data-toc-modified-id="Разделение-данных-на-обучающую,-валидационную-и-тестовую-выборки.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Разделение данных на обучающую, валидационную и тестовую выборки.</a></span></li><li><span><a href="#Модели-машинного-обучения-и-подбор-гиперпараметров" data-toc-modified-id="Модели-машинного-обучения-и-подбор-гиперпараметров-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Модели машинного обучения и подбор гиперпараметров</a></span><ul class="toc-item"><li><span><a href="#DTC-модель-решающего-дерева-с-подбором-гиперпараметров" data-toc-modified-id="DTC-модель-решающего-дерева-с-подбором-гиперпараметров-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>DTC модель решающего дерева с подбором гиперпараметров</a></span></li><li><span><a href="#RFC-модель-случайного-леса-с-подбором-гиперпараметров" data-toc-modified-id="RFC-модель-случайного-леса-с-подбором-гиперпараметров-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>RFC модель случайного леса с подбором гиперпараметров</a></span></li><li><span><a href="#LR-модель-логистической-регрессии" data-toc-modified-id="LR-модель-логистической-регрессии-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>LR модель логистической регрессии</a></span></li><li><span><a href="#Сравнительная-таблица-точности-различных-моделей" data-toc-modified-id="Сравнительная-таблица-точности-различных-моделей-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Сравнительная таблица точности различных моделей</a></span></li><li><span><a href="#Вывод-по-моделям-машинного-обучения" data-toc-modified-id="Вывод-по-моделям-машинного-обучения-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Вывод по моделям машинного обучения</a></span></li></ul></li><li><span><a href="#Проверка-точности-лучшей-модели-на-тестовой-выборке" data-toc-modified-id="Проверка-точности-лучшей-модели-на-тестовой-выборке-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Проверка точности лучшей модели на тестовой выборке</a></span></li><li><span><a href="#Проверка-модели-на-вменяемость-с-помощью-фиктивного-классификатора-DummyClassifier" data-toc-modified-id="Проверка-модели-на-вменяемость-с-помощью-фиктивного-классификатора-DummyClassifier-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Проверка модели на вменяемость с помощью фиктивного классификатора DummyClassifier</a></span><ul class="toc-item"><li><span><a href="#Распределение-пользователей-по-тарифам." data-toc-modified-id="Распределение-пользователей-по-тарифам.-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Распределение пользователей по тарифам.</a></span></li><li><span><a href="#DC-модель-фиктивного-классификатора" data-toc-modified-id="DC-модель-фиктивного-классификатора-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>DC модель фиктивного классификатора</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Введение

### Описание данных
Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

### Инструкция по выполнению проекта

1. Откройте и изучите данные.
2. Разделите исходные данные на обучающую, валидационную и тестовую выборки.
3. Исследуйте качество разных моделей, меняя гиперпараметры. Кратко напишите выводы исследования.
4. Проверьте качество модели на тестовой выборке.
5. Дополнительное задание: проверьте модели на вменяемость.

## Импорт библиотек

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier

# Отключение предупреждений
import warnings
warnings.filterwarnings('ignore')

## Считывание данных

In [2]:
data = pd.read_csv('datasets/users_behavior.csv')
display(data.head())

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

## Разделение данных на обучающую, валидационную и тестовую выборки.

In [3]:
# Создание датафрейма признаки и целевой признаки
features = data.drop(['is_ultra'], axis=1)
target = data['is_ultra']

In [4]:
# Разделение данных на не тестовые 80% и тестовые 20% 
features_notest, features_test, target_notest, target_test = train_test_split(
    features, target, test_size=0.2, random_state=12345
)
# Разделение не тестовой выборки на тренировочную 75% и валидационную 25%
features_train, features_valid, target_train, target_valid = train_test_split(
    features_notest, target_notest, test_size=0.25, random_state=12345
)

## Модели машинного обучения и подбор гиперпараметров

### DTC модель решающего дерева с подбором гиперпараметров

In [5]:
dtc_depth_accuracy = pd.DataFrame(columns=['depth', 'accuracy_train', 'accuracy_valid'])
best_model_dtc = None
best_accuracy_dtc = 0
best_depth_dtc = None

for depth in range(1, 10, 1):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train)
    predictions_train = model.predict(features_train)
    accuracy_train = accuracy_score(target_train, predictions_train)
    accuracy_valid = accuracy_score(target_valid, model.predict(features_valid))
    dtc_depth_accuracy.loc[len(dtc_depth_accuracy.index)] = [depth, accuracy_train, accuracy_valid]
    if accuracy_valid > best_accuracy_dtc:
        best_model_dtc = model
        best_accuracy_dtc = accuracy_valid
        best_depth_dtc = depth
        
display(dtc_depth_accuracy)
print('Точность лучшей модели дерева решений на валидационной выборке', round(best_accuracy_dtc, 2), 
      'Глубина решающего дерева', best_depth_dtc)


depth  accuracy_train  accuracy_valid
0    1.0        0.758299        0.738725
1    2.0        0.792012        0.757387
2    3.0        0.811722        0.765163
3    4.0        0.820539        0.763608
4    5.0        0.827282        0.758942
5    6.0        0.833506        0.757387
6    7.0        0.850622        0.774495
7    8.0        0.866183        0.766719
8    9.0        0.875000        0.762053

Точность лучшей модели дерева решений на валидационной выборке 0.77 Глубина решающего дерева 7


### RFC модель случайного леса с подбором гиперпараметров

In [6]:
rfc_est_accuracy = pd.DataFrame(columns=['estimators', 'max_depth', 'accuracy_train', 'accuracy_valid'])
best_model_rfc = None
best_accuracy_rfc = 0
best_estimators_rfc = 0
best_max_depth = 0

for est in range(1, 20, 1):
    for max_depth in range(1, 10):
        model = RandomForestClassifier(n_estimators=est, max_depth = max_depth, random_state=12345)
        model.fit(features_train, target_train)
        accuracy_train = model.score(features_train, target_train)
        accuracy_valid = model.score(features_valid, target_valid)
        rfc_est_accuracy.loc[len(rfc_est_accuracy.index)] = [est, max_depth, accuracy_train, accuracy_valid]
        if accuracy_valid > best_accuracy_rfc:
            best_model_rfc = model
            best_estimator_rfc = est
            best_max_depth = max_depth
            best_accuracy_rfc = accuracy_valid
                                         
display(rfc_est_accuracy)
print('Точность лучшей модели случайного леса на валидационной выборке', round(best_accuracy_rfc, 2),
      'количество деревьев', best_estimator_rfc,
      'глубина деревьев', max_depth)                                        

estimators  max_depth  accuracy_train  accuracy_valid
0           1.0        1.0        0.746888        0.726283
1           1.0        2.0        0.788900        0.757387
2           1.0        3.0        0.790975        0.757387
3           1.0        4.0        0.797718        0.769829
4           1.0        5.0        0.807573        0.771384
..          ...        ...             ...             ...
166        19.0        5.0        0.831432        0.777605
167        19.0        6.0        0.845954        0.783826
168        19.0        7.0        0.857365        0.786936
169        19.0        8.0        0.866701        0.783826
170        19.0        9.0        0.878112        0.794712

[171 rows x 4 columns]

Точность лучшей модели случайного леса на валидационной выборке 0.8 количество деревьев 3 глубина деревьев 9


### LR модель логистической регрессии

In [7]:
model_lr = LogisticRegression()
model_lr.fit(features_train, target_train)
accuracy_lr = model_lr.score(features_valid, target_valid)
print('Точность линейной регрессии на валидационной выборке', round(accuracy_lr, 2))

Точность линейной регрессии на валидационной выборке 0.7


### Сравнительная таблица точности различных моделей

In [8]:
model_accuracy_valid = pd.DataFrame(columns=['model', 'accuracy_valid'])
model_accuracy_valid.loc[len(model_accuracy_valid.index)] = ['DTC', best_accuracy_dtc]
model_accuracy_valid.loc[len(model_accuracy_valid.index)] = ['RFC', best_accuracy_rfc]
model_accuracy_valid.loc[len(model_accuracy_valid.index)] = ['LR', accuracy_lr]
model_accuracy_valid['accuracy_valid'] = round(model_accuracy_valid['accuracy_valid'], 2)
display(model_accuracy_valid)

model  accuracy_valid
0   DTC            0.77
1   RFC            0.80
2    LR            0.70

### Вывод по моделям машинного обучения

Были построены 3 классификационные модели машинного обучения: дерево решений, случайный лес и логистическая регресия. У 2-х моделей, дерева решений и случайного леса, был проведён подбор гиперпараметров и отобрана лучшая модель с максимальным показателем точности. <br>
Самой лучшей моделью по параметру точности оказалась модель случайого леса с точностью 0,8 с количеством деревьев 3 и глубиной деревьев 9. Второй моделью было дерево решений с точностью 0,77. И модель с самой низкой точностью 0,73 - была логистическая регрессия.

## Проверка точности лучшей модели на тестовой выборке

Проверка лучшей модели случайного леса с количеством деревьев 3 и глубиной деревьев 9 на тестовой выборке.

In [9]:
rfc_accuracy_test = best_model_rfc.score(features_test, target_test)
print('Точность лучшей модели случайного леса на тестовой выбрке:', round(rfc_accuracy_test, 2))

Точность лучшей модели случайного леса на тестовой выбрке: 0.79


## Проверка модели на вменяемость с помощью фиктивного классификатора DummyClassifier

### Распределение пользователей по тарифам.

In [10]:
distribution_tariff_s = data['is_ultra'].value_counts()
distribution_tariff = pd.DataFrame(distribution_tariff_s).reset_index()
distribution_tariff = distribution_tariff.rename(columns={'index' : 'tariff', 'is_ultra' : 'count'})
distribution_tariff.loc[0, 'tariff'] = 'smart'
distribution_tariff.loc[1, 'tariff'] = 'ultra'
distribution_tariff['ratio'] = (
    round(100 * distribution_tariff['count'] / (distribution_tariff.loc[0, 'count'] + distribution_tariff.loc[1, 'count']))
) 
display(distribution_tariff)

tariff  count  ratio
0  smart   2229   69.0
1  ultra    985   31.0

Распределение получилось неравномерное, пользователей тарифа smart - 69%, пользователей тарифа ultra - 31%.

### DC модель фиктивного классификатора

Для проверки моделей на вменяемость воспользуемся фиктивным классификатором DummyClassifier из библиотеки sklearn.dummy. В этой модели фиктивного классификатора посчитаем точность при 2-х разных параметрах strategy, первый, uniform - случайное равномерное распределение целевого признака, и второй, stratified - случайное стратифицированное распределение в соответствии с распределением целевого признака.

In [11]:
#модель DummyClassifier(strategy='uniform)
model_dc_uniform = DummyClassifier(strategy='uniform', random_state=12345)
model_dc_uniform.fit(features_train, target_train)
dc_uniform_accuracy_test = model_dc_uniform.score(features_test, target_test)
print("Ошибка DummyClassifier(strategy='uniform')", round(dc_uniform_accuracy_test, 2))

Ошибка DummyClassifier(strategy='uniform') 0.49


In [12]:
#модель DummyClassifier(strategy='stratified')
model_dc_stratified = DummyClassifier(strategy='stratified', random_state=12345)
model_dc_stratified.fit(features_train, target_train)
dc_stratified_accuracy_test = model_dc_stratified.score(features_test, target_test)
print("Ошибка DummyClassifier(strategy='stratified')", round(dc_stratified_accuracy_test, 2))

Ошибка DummyClassifier(strategy='stratified') 0.58


### Вывод

Проверка моделей на вменяемость показала, что показатель точности лучшей модели случайного леса - 0.8, выше показателей точности моделей со случайной классификацией 0.58. Таким образом, модели можно использовать для предсказания тарифов.

## Общий вывод

Проведенная работа по построению, подбору гиперпараметров и отбору моделей машинного обучения для предсказания тарифов по активности пользователя позволяет рекомендовать для использования модель случайного леса с количеством деревьев 3 и глубиной 9. Точность модели случайного леса на тествовой выборке 0,79. 